diff --git a/dist/index.js b/dist/index.js
index ddc9b66d15374bee857b2f62b02098ef09d0bc36..e7c8cc6c7eacc3e63c1be1e8a3ddf1970cab0e45 100644
--- a/dist/index.js
+++ b/dist/index.js
@@ -139,7 +139,7 @@ async function decryptWithKey(encryptionKey, payload) {
     const key = unwrapKey(encryptionKey);
     let decryptedObj;
     try {
-        const result = await crypto.subtle.decrypt({ name: DERIVED_KEY_FORMAT, iv: vector }, key, encryptedData);
+        const result = await globalThis.crypto.subtle.decrypt({ name: DERIVED_KEY_FORMAT, iv: vector }, key, encryptedData);
         const decryptedData = new Uint8Array(result);
         const decryptedStr = Buffer.from(decryptedData).toString(STRING_ENCODING);
         decryptedObj = JSON.parse(decryptedStr);
@@ -193,12 +193,18 @@ async function keyFromPassword(password, salt, exportable = false, opts = OLD_DE
     const passBuffer = Buffer.from(password, STRING_ENCODING);
     const saltBuffer = Buffer.from(salt, 'base64');
     const key = await globalThis.crypto.subtle.importKey('raw', passBuffer, { name: 'PBKDF2' }, false, ['deriveBits', 'deriveKey']);
-    const derivedKey = await globalThis.crypto.subtle.deriveKey({
-        name: 'PBKDF2',
-        salt: saltBuffer,
-        iterations: opts.params.iterations,
-        hash: 'SHA-256',
-    }, key, { name: DERIVED_KEY_FORMAT, length: 256 }, exportable, ['encrypt', 'decrypt']);
+    const derivedBits = await globalThis.crypto.subtle.deriveBits(
+        { name: 'PBKDF2', salt: saltBuffer, iterations: opts.params.iterations, hash: 'SHA-256' },
+        key,
+        256
+    );
+    const derivedKey = await globalThis.crypto.subtle.importKey(
+        'raw', 
+        derivedBits,
+        { name: DERIVED_KEY_FORMAT, length: 256 },
+        exportable,
+        ['encrypt', 'decrypt']
+    );
     return opts
         ? {
             key: derivedKey,
@@ -317,7 +323,7 @@ function isEncryptionKey(encryptionKey) {
     return ((0, utils_1.isPlainObject)(encryptionKey) &&
         (0, utils_1.hasProperty)(encryptionKey, 'key') &&
         (0, utils_1.hasProperty)(encryptionKey, 'derivationOptions') &&
-        encryptionKey.key instanceof CryptoKey &&
+        encryptionKey.key instanceof globalThis.crypto.CryptoKey &&
         isKeyDerivationOptions(encryptionKey.derivationOptions));
 }
 /**
