diff --git a/android/build.gradle b/android/build.gradle
index 45e323e8fe045f2d9c74c912a446cfc81ce35b6b..463c0e1c95f10540eac9a290a21a33bf355e0358 100644
--- a/android/build.gradle
+++ b/android/build.gradle
@@ -112,7 +112,9 @@ android {
     externalNativeBuild {
       cmake {
         cppFlags "-O2 -frtti -fexceptions -Wall -Wno-unused-variable -fstack-protector-all -DANDROID"
-        arguments "-DANDROID_STL=c++_shared"
+        arguments "-DANDROID_STL=c++_shared",
+                  "-DANDROID_SUPPORT_FLEXIBLE_PAGE_SIZES=ON", // don't need patch this after upgrade to 0.7.17
+                  "-DNODE_MODULES_DIR=${nodeModules}", "-DRABBY_QUICK_CRYPTO_BUILD_TYPE=${resolveBuildType()}"
         abiFilters (*reactNativeArchitectures())
       }
     }
@@ -175,7 +177,7 @@ dependencies {
   implementation "com.facebook.react:react-android:+"
 
   // Add a dependency on OpenSSL
-  implementation 'io.github.ronickg:openssl:3.3.2'
+  implementation 'io.github.ronickg:openssl:3.3.2-1' // don't need patch this after upgrade to 0.7.17
 }
 
 // Resolves "LOCAL_SRC_FILES points to a missing file, Check that libfb.so exists or that its path is correct".
diff --git a/lib/module/index.js b/lib/module/index.js
index f52cc589a4aef892833ecb2c43a547927c435a3d..8f5e1575519fcbbfed7e35ecc3bdfa1a2743c84b 100644
--- a/lib/module/index.js
+++ b/lib/module/index.js
@@ -12,7 +12,7 @@ import { constants } from './constants';
 import { subtle } from './subtle';
 import { getCiphers, getHashes } from './Utils';
 import webcrypto from './webcrypto';
-import { createPrivateKey, createPublicKey, createSecretKey } from './keys';
+import * as keys from './keys';
 
 /**
  * Loosely matches Node.js {crypto} with some unimplemented functionality
@@ -26,9 +26,6 @@ const QuickCrypto = {
   createCipheriv,
   createDecipher,
   createDecipheriv,
-  createPublicKey,
-  createPrivateKey,
-  createSecretKey,
   publicEncrypt,
   publicDecrypt,
   privateDecrypt,
@@ -44,7 +41,8 @@ const QuickCrypto = {
   ...random,
   getCiphers,
   getHashes,
-  webcrypto
+  webcrypto,
+  ...keys
 };
 
 /**
