diff --git a/android/src/main/java/com/mrousavy/camera/react/CameraDevicesManager.kt b/android/src/main/java/com/mrousavy/camera/react/CameraDevicesManager.kt
index 21ac2b284217f21dbbc9c2489ca8ccc358438f31..03864788d82205627e84362690582a0fff1ec344 100644
--- a/android/src/main/java/com/mrousavy/camera/react/CameraDevicesManager.kt
+++ b/android/src/main/java/com/mrousavy/camera/react/CameraDevicesManager.kt
@@ -71,7 +71,6 @@ class CameraDevicesManager(private val reactContext: ReactApplicationContext) :
       } catch (error: Throwable) {
         Log.e(TAG, "Failed to initialize ProcessCameraProvider/ExtensionsManager! Error: ${error.message}", error)
       }
-      sendAvailableDevicesChangedEvent()
     }
   }
 
@@ -79,6 +78,7 @@ class CameraDevicesManager(private val reactContext: ReactApplicationContext) :
   override fun initialize() {
     super.initialize()
     cameraManager.registerAvailabilityCallback(callback, null)
+    sendAvailableDevicesChangedEvent()
   }
 
   override fun invalidate() {
